File: sdk/android/src/main/java/com/tuya/smart/rnsdk/home/TuyaHomeManagerModule.java
package com.tuya.smart.rnsdk.home;

import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.tuya.smart.home.sdk.TuyaHomeSdk;
import com.tuya.smart.home.sdk.api.ITuyaHomeChangeListener;
import com.tuya.smart.home.sdk.bean.HomeBean;
import com.tuya.smart.home.sdk.callback.ITuyaGetHomeListCallback;
import com.tuya.smart.home.sdk.callback.ITuyaHomeResultCallback;
import com.tuya.smart.rnsdk.utils.BridgeUtils;
import com.tuya.smart.rnsdk.utils.Constant;
import com.tuya.smart.rnsdk.utils.JsonUtils;
import com.tuya.smart.rnsdk.utils.ReactParamsCheck;
import com.tuya.smart.rnsdk.utils.TuyaReactUtils;
import com.tuya.smart.sdk.bean.DeviceBean;
import com.tuya.smart.sdk.bean.GroupBean;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.annotation.Nonnull;

import static com.tuya.smart.rnsdk.utils.Constant.GEONAME;
import static com.tuya.smart.rnsdk.utils.Constant.LAT;
import static com.tuya.smart.rnsdk.utils.Constant.LON;
import static com.tuya.smart.rnsdk.utils.Constant.NAME;
import static com.tuya.smart.rnsdk.utils.Constant.ROMMS;

public class TuyaHomeManagerModule extends ReactContextBaseJavaModule {
    public TuyaHomeManagerModule(@Nonnull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    private ITuyaHomeChangeListener mITuyaHomeChangeListener;

    @Nonnull
    @Override
    public String getName() {
        return "TuyaHomeManagerModule";
    }

    @ReactMethod
    public void queryHomeList(final Promise promise) {
        TuyaHomeSdk.getHomeManagerInstance().queryHomeList(new ITuyaGetHomeListCallback() {
            @Override
            public void onSuccess(List<HomeBean> homeBeans) {
                promise.resolve(TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(homeBeans)));
            }

            @Override
            public void onError(String errorCode, String error) {
                promise.reject(errorCode, error);
            }
        });
    }

    @ReactMethod
    public void createHome(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(NAME, LON, LAT, GEONAME, ROMMS), params)) {
            ArrayList<String> list = new ArrayList<>();
            for (int i = 0; i < params.getArray(ROMMS).size(); i++) {
                list.add(params.getArray(ROMMS).getString(i));
            }
            TuyaHomeSdk.getHomeManagerInstance().createHome(
                    params.getString(Constant.NAME),
                    params.getDouble(Constant.LON),
                    params.getDouble(Constant.LAT),
                    params.getString(Constant.GEONAME),
                    list,
                    new ITuyaHomeResultCallback() {
                        @Override
                        public void onSuccess(HomeBean bean) {
                            promise.resolve(TuyaReactUtils.parseToWritableMap(bean));
                        }

                        @Override
                        public void onError(String errorCode, String errorMsg) {
                            promise.reject(errorCode, errorMsg);
                        }
                    }
            );
        }
    }


    @ReactMethod
    public void registerTuyaHomeChangeListener() {
        if (mITuyaHomeChangeListener != null) {
            TuyaHomeSdk.getHomeManagerInstance().unRegisterTuyaHomeChangeListener(mITuyaHomeChangeListener);
            mITuyaHomeChangeListener = null;
        }
        mITuyaHomeChangeListener = new ITuyaHomeChangeListener() {

            @Override
            public void onHomeAdded(long homeId) {
                WritableMap map = Arguments.createMap();
                map.putDouble("homeId", homeId);
                map.putString("type", "onHomeAdded");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }

            @Override
            public void onHomeInvite(long homeId, String homeName) {
                WritableMap map = Arguments.createMap();
                map.putDouble("homeId", homeId);
                map.putString("homeName", homeName);
                map.putString("type", "onHomeInvite");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }

            @Override
            public void onHomeRemoved(long homeId) {
                WritableMap map = Arguments.createMap();
                map.putDouble("homeId", homeId);
                map.putString("type", "onHomeRemoved");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }

            @Override
            public void onHomeInfoChanged(long homeId) {
                WritableMap map = Arguments.createMap();
                map.putDouble("homeId", homeId);
                map.putString("type", "onHomeInfoChanged");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }

            @Override
            public void onSharedDeviceList(List<DeviceBean> sharedDeviceList) {
                WritableMap map = Arguments.createMap();
                map.putArray("sharedDeviceList", TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(sharedDeviceList)));
                map.putString("type", "onSharedDeviceList");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }

            @Override
            public void onSharedGroupList(List<GroupBean> sharedGroupList) {
                WritableMap map = Arguments.createMap();
                map.putArray("sharedGroupList", TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(sharedGroupList)));
                map.putString("type", "onSharedGroupList");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }

            @Override
            public void onServerConnectSuccess() {
                WritableMap map = Arguments.createMap();
                map.putString("type", "onServerConnectSuccess");
                BridgeUtils.homeChange(getReactApplicationContext(), map, "");
            }
        };
        TuyaHomeSdk.getHomeManagerInstance().registerTuyaHomeChangeListener(mITuyaHomeChangeListener);
    }

    @ReactMethod
    public void unregisterTuyaHomeChangeListener() {
        TuyaHomeSdk.getHomeManagerInstance().unRegisterTuyaHomeChangeListener(mITuyaHomeChangeListener);
        mITuyaHomeChangeListener = null;
    }


    @ReactMethod
    public void onDestroy() {
        TuyaHomeSdk.getHomeManagerInstance().onDestroy();
    }
}

File: Example/android/tuyasmart-sdk-react-native/src/main/java/com/tuya/smart/rnsdk/home/TuyaHomeManagerModule.kt
package com.tuya.smart.rnsdk.home

import com.facebook.react.bridge.*
import com.tuya.smart.home.sdk.TuyaHomeSdk
import com.tuya.smart.home.sdk.api.ITuyaHomeChangeListener
import com.tuya.smart.home.sdk.bean.HomeBean
import com.tuya.smart.home.sdk.callback.ITuyaGetHomeListCallback
import com.tuya.smart.home.sdk.callback.ITuyaHomeResultCallback
import com.tuya.smart.rnsdk.utils.*
import com.tuya.smart.sdk.bean.DeviceBean
import com.tuya.smart.sdk.bean.GroupBean

class TuyaHomeManagerModule(reactContext: ReactApplicationContext?) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String {
        return "TuyaHomeManagerModule"
    }


    /* 获取家庭列表 */
    @ReactMethod
    fun queryHomeList(promise: Promise) {
        TuyaHomeSdk.getHomeManagerInstance().queryHomeList(object : ITuyaGetHomeListCallback {
            override fun onSuccess(var1: List<HomeBean>) {
                promise.resolve(TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(var1!!)))
            }

            override fun onError(var1: String, var2: String) {
                promise.reject(var1, var2)
            }
        })
    }

    /* 创建家庭 */
    @ReactMethod
    fun createHome(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(Constant.NAME, Constant.LON, Constant.LAT, Constant.GEONAME, Constant.ROMMS), params)) {
            var list = ArrayList<String>()
            var length = params.getArray(Constant.ROMMS).size()
            for (index in 0 until length) {
                list.add(params.getArray(Constant.ROMMS).getString(index))
            }
            TuyaHomeSdk.getHomeManagerInstance().createHome(
                    params.getString(Constant.NAME),
                    params.getDouble(Constant.LON),
                    params.getDouble(Constant.LAT),
                    params.getString(Constant.GEONAME),
                    list,
                    getITuyaHomeResultCallback(promise)
            )

        }
    }


    /* 注册家庭信息的变更
     * 有：家庭的增加、删除、信息变更、分享列表的变更和服务器连接成功的监听 */
    @ReactMethod
    fun registerTuyaHomeChangeListener(params: ReadableMap) {
        TuyaHomeSdk.getHomeManagerInstance().registerTuyaHomeChangeListener(object : ITuyaHomeChangeListener {
            override fun onHomeAdded(var1: Long) {
                val map = Arguments.createMap()
                map.putDouble("homeId", var1.toDouble())
                map.putString("type", "onHomeAdded");
                BridgeUtils.homeChange(reactApplicationContext, map, params.getString("homeId"))
            }

            override fun onHomeRemoved(var1: Long) {
                val map = Arguments.createMap()
                map.putDouble("homeId", var1.toDouble())
                map.putString("type", "onHomeRemoved");
                BridgeUtils.homeChange(reactApplicationContext, map, params.getString("homeId"))
            }

            override fun onHomeInfoChanged(var1: Long) {
                val map = Arguments.createMap()
                map.putDouble("homeId", var1.toDouble())
                map.putString("type", "onHomeInfoChanged");
                BridgeUtils.homeChange(reactApplicationContext, map, params.getString("homeId"))
            }

            override fun onSharedDeviceList(var1: List<DeviceBean>) {
                val map = Arguments.createMap()
                map.putArray("deviceBeans", TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(var1)))
                map.putString("type", "onSharedDeviceList");
                BridgeUtils.homeChange(reactApplicationContext, map, params.getString("homeId"))
            }

            override fun onSharedGroupList(var1: List<GroupBean>) {
                val map = Arguments.createMap()
                map.putArray("groupBeans", TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(var1)))
                map.putString("type", "onSharedGroupList");
                BridgeUtils.homeChange(reactApplicationContext, map, params.getString("homeId"))
            }

            override fun onServerConnectSuccess() {
                val map = Arguments.createMap()
                map.putString("type", "onServerConnectSuccess");
                BridgeUtils.homeChange(reactApplicationContext, map, params.getDouble("homeId").toString())
            }
        })
    }


    fun getITuyaHomeResultCallback(promise: Promise): ITuyaHomeResultCallback? {
        return object : ITuyaHomeResultCallback {
            override fun onSuccess(p0: HomeBean?) {
                promise.resolve(TuyaReactUtils.parseToWritableMap(p0))
            }

            override fun onError(code: String?, error: String?) {
                promise.reject(code, error)
            }
        }
    }
}